### PR TITLE
fix(langchain): detach orphaned context_api.attach() calls that corrupt OTel context

### LINKED ISSUE #3526: 🐛 Bug Report: Orphaned `context_api.attach()` Calls Corrupt OpenTelemetry Context Stack
### Which component is this bug for?

Langchain Instrumentation

### 📜 Description

The LangChain instrumentation library has multiple code paths where `context_api.attach()` is called without a corresponding `context_api.detach()`, leaving orphaned contexts on the stack. This corrupts the OpenTelemetry context, causing subsequent code to lose access to the active trace context.

### Impact

After LangChain/LangGraph execution completes, `trace.get_current_span()` returns an **ended span** (with `is_recording() == False`) instead of the parent span that should be active. This causes:

1. **Missing trace IDs in logs**: Logging systems that check `span.is_recording()` before adding trace context see `False` and skip adding `trace_id`/`span_id`
2. **Broken span hierarchy**: Child spans created after LangChain execution may attach to the wrong parent
3. **Inconsistent tracing**: The same application has some logs with trace context and others without

### Locations of Orphaned Attaches

**File:** `opentelemetry/instrumentation/langchain/callback_handler.py`

#### 1. `_create_span()` method (lines 263-269)
```python
if metadata is not None:
    current_association_properties = (
        context_api.get_value("association_properties") or {}
    )
    try:
        context_api.attach(  # ❌ No detach, token not saved
            context_api.set_value(
                "association_properties",
                {**current_association_properties, **sanitized_metadata},
            )
        )
    except Exception:
        pass
```

#### 2. `on_chain_end()` method (lines 462-471)
```python
self._end_span(span, run_id)
if parent_run_id is None:
    try:
        context_api.attach(  # ❌ No detach, token not saved
            context_api.set_value(
                SUPPRESS_LANGUAGE_MODEL_INSTRUMENTATION_KEY, False
            )
        )
    except Exception:
        pass
```

### Workaround We're Using

We implemented a context manager that saves the OpenTelemetry context before calling LangChain/LangGraph and restores it afterward:

```python
from contextlib import contextmanager
from collections.abc import Generator
from opentelemetry import context as otel_context

@contextmanager
def preserve_otel_context() -> Generator[None]:
    """
    Preserve OpenTelemetry context across operations that may detach it.
    """
    token = otel_context.attach(otel_context.get_current())
    try:
        yield
    finally:
        otel_context.detach(token)
```

Usage:
```python
with preserve_otel_context():
    result = await agent.ainvoke(...)
# Context is now correctly restored
```

This workaround doesn't fix the library's context leaks but allows us to restore the correct parent span after execution.

### 👟 Reproduction steps

1. **Set up OpenTelemetry with LangChain instrumentation:**
```python
from opentelemetry import trace
from opentelemetry.instrumentation.langchain import LangchainInstrumentor

LangchainInstrumentor().instrument()
tracer = trace.get_tracer(__name__)
```

2. **Create a parent span and invoke a LangChain chain inside it:**
```python
with tracer.start_as_current_span("parent_operation"):
    # Before LangChain
    span_before = trace.get_current_span()
    print(f"Before: {span_before.name}, recording={span_before.is_recording()}")

    # Invoke any LangChain chain or LangGraph
    result = await chain.ainvoke(input_data)

    # After LangChain
    span_after = trace.get_current_span()
    print(f"After: {span_after.name}, recording={span_after.is_recording()}")
```

3. **Observe the output:**
```
Before: parent_operation, recording=True
After: some_langchain_span, recording=False  # ❌ Wrong span, already ended
```

### 👍 Expected behavior

After LangChain/LangGraph execution completes:
- `trace.get_current_span()` should return the **parent span** (`parent_operation`)
- The parent span should have `is_recording() == True`
- Subsequent logging or span creation should use the correct parent context

### 👎 Actual Behavior with Screenshots

_FYI I am including logs instead of screenshots._

After LangChain/LangGraph execution, the context points to an **ended LangChain instrumentation span** instead of the parent:

```
[TRACE DEBUG] Before ainvoke - Span: _Span(name="case_worker.process", context=SpanContext(trace_id=0x1c60dc422d09a853510b4c8f688decc7, span_id=0x7349095e5220669a, trace_flags=0x01, trace_state=[], is_remote=False)), Recording: True, Trace ID: 1c60dc422d09a853510b4c8f688decc7

[TRACE DEBUG] Inside preserve_otel_context, before ainvoke - Span: _Span(name="case_worker.process", context=SpanContext(trace_id=0x1c60dc422d09a853510b4c8f688decc7, span_id=0x7349095e5220669a, trace_flags=0x01, trace_state=[], is_remote=False)), Recording: True, Trace ID: 1c60dc422d09a853510b4c8f688decc7

# After LangGraph ainvoke() returns:
[TRACE DEBUG] Inside preserve_otel_context, after ainvoke - Span: _Span(name="judge.task", context=SpanContext(trace_id=0x1c60dc422d09a853510b4c8f688decc7, span_id=0x12ed093291f1de81, trace_flags=0x01, trace_state=[], is_remote=False)), Recording: False, Trace ID: None

# With our workaround, context is restored:
[TRACE DEBUG] After exiting preserve_otel_context - Span: _Span(name="case_worker.process", context=SpanContext(trace_id=0x1c60dc422d09a853510b4c8f688decc7, span_id=0x7349095e5220669a, trace_flags=0x01, trace_state=[], is_remote=False)), Recording: True, Trace ID: 1c60dc422d09a853510b4c8f688decc7
```

**Key observations:**
1. After `ainvoke()`, the current span changed from `case_worker.process` to `judge.task`
2. The `judge.task` span has `Recording: False` (it was ended)
3. Without our workaround, subsequent logs would have no trace_id because the span is not recording

### 🤖 Python Version

3.13.7

### 📃 Provide any additional context for the Bug.

# Versions
- **opentelemetry-instrumentation-langchain:** 0.50.1
- **opentelemetry-api:** 1.29.0
- **langchain-core:** 0.3.29
- **langgraph:** 0.2.63

### 👀 Have you spent some time to check if this bug has been raised before?

- [x] I checked and didn't find similar issue

### Are you willing to submit PR?

Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by saivedant169**
Fixes #3526

## Description

The LangChain instrumentation has two code paths where `context_api.attach()` is called without a corresponding `context_api.detach()`, leaving orphaned contexts on the OpenTelemetry stack. After LangChain execution completes, `trace.get_current_span()` returns an ended span instead of the parent span, breaking downstream trace context propagation and log correlation.

### 1. `_create_span()` — association_properties token lost

```python
# BEFORE: token never saved, never detached
context_api.attach(
    context_api.set_value("association_properties", {...})
)
```

**Fix:** Save the token and store it in a new `SpanHolder.association_properties_token` field. Detach it in `_end_span()` alongside the span token.

### 2. `on_chain_end()` — redundant attach without detach

```python
# BEFORE: orphaned attach to reset suppression key
context_api.attach(
    context_api.set_value(SUPPRESS_LANGUAGE_MODEL_INSTRUMENTATION_KEY, False)
)
```

**Fix:** Removed entirely. The suppression token is already properly detached by `_end_span()` via `SpanHolder.token`, so this second attach was both redundant and harmful.

## Changes

- `span_utils.py`: Added `association_properties_token` field to `SpanHolder` dataclass
- `callback_handler.py`: Save association_properties attach token and detach in `_end_span()`
- `callback_handler.py`: Remove redundant orphaned attach in `on_chain_end()`

## Testing

- `ruff check` and `ruff format` pass on all changed files
- The first commit is a pre-existing formatting fix (ruff format applied to the two files); the second commit contains the actual bug fix

## Summary by CodeRabbit

* **Refactor**
  * Improved span lifecycle and context handling for more reliable instrumentation
  * Added management of association properties to better preserve span-associated state
  * Safer teardown/detachment of span-related context to reduce leaks or misattribution
  * Removed an unnecessary conditional re-attachment step to simplify end-of-chain behavior

### CODE REVIEW

**Comment by saivedant169** [comment]
Thanks for the thorough analysis @coderabbitai. 

The point about `_create_llm_span()` overwriting the SpanHolder (line 464) is valid, but it's **pre-existing behavior** that was in the code before this PR — my changes only fix the two orphaned `context_api.attach()` calls reported in #3526. The `_create_llm_span` overwrite issue is a separate concern and should be tracked in its own issue if maintainers want it addressed.

The Sourcery review approved the changes. Happy to address any maintainer feedback.

**Comment by saivedant169** [comment]
yeah that'd be helpful @CodeRabbit , go ahead and open one. thanks

**Comment by saivedant169** [comment]
Closing this. The orphaned context_api.attach() cleanup landed on main via a different change that wraps the same logic in a _detach_holder_contexts helper and calls it from both _end_span (for the run and its children) and _create_span (when an existing holder is overwritten). Same cases handled, just with the helper-based pattern. Tracking issue #3957 is also addressed there. Thanks for the back-and-forth on this earlier.
